File: go/mechanisms/evm/erc4337_networks.go
package evm

import (
	"fmt"
	"strconv"
	"strings"
)

// ERC4337ChainInfo contains information about an ERC-4337 supported chain.
type ERC4337ChainInfo struct {
	ChainID                    int
	Name                       string
	V1Name                     string
	CAIP2                      string
	RpcUrl                     string
	BlockExplorerUrl           string
	UsdcAddress                string
	SafeTransactionServiceUrl  string
	Testnet                    bool
}

// ERC4337SupportedChains maps chain IDs to their chain info.
var ERC4337SupportedChains = map[int]*ERC4337ChainInfo{
	8453: {
		ChainID:                   8453,
		Name:                      "Base",
		V1Name:                    "base",
		CAIP2:                     "eip155:8453",
		RpcUrl:                    "https://mainnet.base.org",
		BlockExplorerUrl:          "https://basescan.org",
		UsdcAddress:               "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
		SafeTransactionServiceUrl: "https://safe-transaction-base.safe.global",
		Testnet:                   false,
	},
	84532: {
		ChainID:                   84532,
		Name:                      "Base Sepolia",
		V1Name:                    "base-sepolia",
		CAIP2:                     "eip155:84532",
		RpcUrl:                    "https://sepolia.base.org",
		BlockExplorerUrl:          "https://sepolia.basescan.org",
		UsdcAddress:               "0x036CbD53842c5426634e7929541eC2318f3dCF7e",
		SafeTransactionServiceUrl: "https://safe-transaction-base-sepolia.safe.global",
		Testnet:                   true,
	},
	10: {
		ChainID:                   10,
		Name:                      "Optimism",
		V1Name:                    "optimism",
		CAIP2:                     "eip155:10",
		RpcUrl:                    "https://mainnet.optimism.io",
		BlockExplorerUrl:          "https://optimistic.etherscan.io",
		UsdcAddress:               "0x0b2C639c533813f4Aa9D7837CAf62653d097Ff85",
		SafeTransactionServiceUrl: "https://safe-transaction-optimism.safe.global",
		Testnet:                   false,
	},
	11155420: {
		ChainID:                   11155420,
		Name:                      "Optimism Sepolia",
		V1Name:                    "optimism-sepolia",
		CAIP2:                     "eip155:11155420",
		RpcUrl:                    "https://sepolia.optimism.io",
		BlockExplorerUrl:          "https://sepolia-optimistic.etherscan.io",
		UsdcAddress:               "0x5fd84259d66Cd46123540766Be93DFE6D43130D7",
		SafeTransactionServiceUrl: "",
		Testnet:                   true,
	},
	42161: {
		ChainID:                   42161,
		Name:                      "Arbitrum One",
		V1Name:                    "arbitrum",
		CAIP2:                     "eip155:42161",
		RpcUrl:                    "https://arb1.arbitrum.io/rpc",
		BlockExplorerUrl:          "https://arbiscan.io",
		UsdcAddress:               "0xaf88d065e77c8cC2239327C5EDb3A432268e5831",
		SafeTransactionServiceUrl: "https://safe-transaction-arbitrum.safe.global",
		Testnet:                   false,
	},
	421614: {
		ChainID:                   421614,
		Name:                      "Arbitrum Sepolia",
		V1Name:                    "arbitrum-sepolia",
		CAIP2:                     "eip155:421614",
		RpcUrl:                    "https://sepolia-rollup.arbitrum.io/rpc",
		BlockExplorerUrl:          "https://sepolia.arbiscan.io",
		UsdcAddress:               "0x75faf114eafb1BDbe2F0316DF893fd58CE46AA4d",
		SafeTransactionServiceUrl: "",
		Testnet:                   true,
	},
}

// erc4337V1NameIndex maps v1 names to chain info.
var erc4337V1NameIndex map[string]*ERC4337ChainInfo

func init() {
	erc4337V1NameIndex = make(map[string]*ERC4337ChainInfo, len(ERC4337SupportedChains))
	for _, chain := range ERC4337SupportedChains {
		erc4337V1NameIndex[chain.V1Name] = chain
	}
}

// GetERC4337Chain returns the chain info for a given chain ID.
// Returns nil if the chain is not supported.
func GetERC4337Chain(chainID int) *ERC4337ChainInfo {
	return ERC4337SupportedChains[chainID]
}

// IsERC4337Supported checks if a chain ID is in the ERC-4337 supported chains registry.
func IsERC4337Supported(chainID int) bool {
	_, ok := ERC4337SupportedChains[chainID]
	return ok
}

// ResolveERC4337ChainId resolves a network input (CAIP-2 string, v1 name, or numeric string)
// to a numeric chain ID in the ERC-4337 registry.
func ResolveERC4337ChainId(network string) (int, error) {
	// Try CAIP-2 format (eip155:CHAIN_ID)
	if strings.HasPrefix(network, "eip155:") {
		parts := strings.SplitN(network, ":", 2)
		if len(parts) != 2 {
			return 0, fmt.Errorf("invalid CAIP-2 identifier: %s", network)
		}
		chainID, err := strconv.Atoi(parts[1])
		if err != nil {
			return 0, fmt.Errorf("invalid CAIP-2 chain ID: %s", network)
		}
		return chainID, nil
	}

	// Try v1 name
	if chain, ok := erc4337V1NameIndex[network]; ok {
		return chain.ChainID, nil
	}

	// Try numeric
	chainID, err := strconv.Atoi(network)
	if err != nil {
		return 0, fmt.Errorf("unknown network: %s. Expected CAIP-2 (eip155:chainId), a known v1 name, or a numeric chain ID", network)
	}
	return chainID, nil
}

File: python/x402/mechanisms/evm/erc4337_networks.py
"""ERC-4337 network registry for x402 EVM mechanism."""

from dataclasses import dataclass


@dataclass
class ERC4337ChainInfo:
    """Information about an ERC-4337 supported chain."""

    chain_id: int
    name: str
    v1_name: str
    caip2: str
    rpc_url: str
    block_explorer_url: str
    usdc_address: str
    safe_transaction_service_url: str | None
    testnet: bool


ERC4337_SUPPORTED_CHAINS: dict[int, ERC4337ChainInfo] = {
    8453: ERC4337ChainInfo(
        chain_id=8453,
        name="Base",
        v1_name="base",
        caip2="eip155:8453",
        rpc_url="https://mainnet.base.org",
        block_explorer_url="https://basescan.org",
        usdc_address="0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
        safe_transaction_service_url="https://safe-transaction-base.safe.global",
        testnet=False,
    ),
    84532: ERC4337ChainInfo(
        chain_id=84532,
        name="Base Sepolia",
        v1_name="base-sepolia",
        caip2="eip155:84532",
        rpc_url="https://sepolia.base.org",
        block_explorer_url="https://sepolia.basescan.org",
        usdc_address="0x036CbD53842c5426634e7929541eC2318f3dCF7e",
        safe_transaction_service_url="https://safe-transaction-base-sepolia.safe.global",
        testnet=True,
    ),
    10: ERC4337ChainInfo(
        chain_id=10,
        name="Optimism",
        v1_name="optimism",
        caip2="eip155:10",
        rpc_url="https://mainnet.optimism.io",
        block_explorer_url="https://optimistic.etherscan.io",
        usdc_address="0x0b2C639c533813f4Aa9D7837CAf62653d097Ff85",
        safe_transaction_service_url="https://safe-transaction-optimism.safe.global",
        testnet=False,
    ),
    11155420: ERC4337ChainInfo(
        chain_id=11155420,
        name="Optimism Sepolia",
        v1_name="optimism-sepolia",
        caip2="eip155:11155420",
        rpc_url="https://sepolia.optimism.io",
        block_explorer_url="https://sepolia-optimistic.etherscan.io",
        usdc_address="0x5fd84259d66Cd46123540766Be93DFE6D43130D7",
        safe_transaction_service_url=None,
        testnet=True,
    ),
    42161: ERC4337ChainInfo(
        chain_id=42161,
        name="Arbitrum One",
        v1_name="arbitrum",
        caip2="eip155:42161",
        rpc_url="https://arb1.arbitrum.io/rpc",
        block_explorer_url="https://arbiscan.io",
        usdc_address="0xaf88d065e77c8cC2239327C5EDb3A432268e5831",
        safe_transaction_service_url="https://safe-transaction-arbitrum.safe.global",
        testnet=False,
    ),
    421614: ERC4337ChainInfo(
        chain_id=421614,
        name="Arbitrum Sepolia",
        v1_name="arbitrum-sepolia",
        caip2="eip155:421614",
        rpc_url="https://sepolia-rollup.arbitrum.io/rpc",
        block_explorer_url="https://sepolia.arbiscan.io",
        usdc_address="0x75faf114eafb1BDbe2F0316DF893fd58CE46AA4d",
        safe_transaction_service_url=None,
        testnet=True,
    ),
}

# V1 name index for reverse lookup
_V1_NAME_INDEX: dict[str, ERC4337ChainInfo] = {
    chain.v1_name: chain for chain in ERC4337_SUPPORTED_CHAINS.values()
}


def get_erc4337_chain(chain_id: int) -> ERC4337ChainInfo | None:
    """Get chain info by chain ID.

    Args:
        chain_id: The numeric chain ID.

    Returns:
        The chain info or None if not found.
    """
    return ERC4337_SUPPORTED_CHAINS.get(chain_id)


def is_erc4337_supported(chain_id: int) -> bool:
    """Check if a chain ID is in the ERC-4337 supported chains registry.

    Args:
        chain_id: The numeric chain ID.

    Returns:
        Whether the chain is supported.
    """
    return chain_id in ERC4337_SUPPORTED_CHAINS


def resolve_erc4337_chain_id(network: str) -> int:
    """Resolve a network input to a numeric chain ID.

    Handles CAIP-2 format (eip155:chainId), v1 names, and numeric strings.

    Args:
        network: The network identifier to resolve.

    Returns:
        The numeric chain ID.

    Raises:
        ValueError: If the network cannot be resolved.
    """
    # Try CAIP-2 format
    if network.startswith("eip155:"):
        parts = network.split(":", 1)
        if len(parts) != 2:
            raise ValueError(f"Invalid CAIP-2 identifier: {network}")
        try:
            return int(parts[1])
        except ValueError:
            raise ValueError(f"Invalid CAIP-2 chain ID: {network}") from None

    # Try v1 name
    if network in _V1_NAME_INDEX:
        return _V1_NAME_INDEX[network].chain_id

    # Try numeric
    try:
        return int(network)
    except ValueError:
        raise ValueError(
            f"Unknown network: {network}. Expected CAIP-2 (eip155:chainId), "
            f"a known v1 name, or a numeric chain ID."
        ) from None


def get_supported_chains() -> list[ERC4337ChainInfo]:
    """Return all supported chains."""
    return list(ERC4337_SUPPORTED_CHAINS.values())


def get_mainnets() -> list[ERC4337ChainInfo]:
    """Return all supported mainnet chains."""
    return [c for c in ERC4337_SUPPORTED_CHAINS.values() if not c.testnet]


def get_testnets() -> list[ERC4337ChainInfo]:
    """Return all supported testnet chains."""
    return [c for c in ERC4337_SUPPORTED_CHAINS.values() if c.testnet]
